File: benchmarks/benchmarks/03_data_structures/comparisons/ring_buffer.go
package main

import "fmt"

func ringBufferCost(capacity int64, rounds int64) int64 {
	var head int64
	var tail int64
	var acc int64
	for item := int64(0); item < rounds; item++ {
		slot := (head + item) % capacity
		acc += ((slot * 17) + item) % 257
		if (item % 3) == 0 {
			tail = (tail + 1) % capacity
			acc += tail
		}
		head = (head + 1) % capacity
	}
	return acc + head + tail
}

func main() {
	fmt.Println(ringBufferCost(4096, 12000000))
}

File: benchmarks/benchmarks/03_data_structures/comparisons/ring_buffer.py
def ring_buffer_cost(capacity: int, rounds: int) -> int:
    head = 0
    tail = 0
    acc = 0
    for item in range(rounds):
        slot = (head + item) % capacity
        acc += ((slot * 17) + item) % 257
        if (item % 3) == 0:
            tail = (tail + 1) % capacity
            acc += tail
        head = (head + 1) % capacity
    return acc + head + tail


if __name__ == "__main__":
    print(ring_buffer_cost(4096, 12000000))
